core/src/main/scala/mongo4cats/helpers.scala

```diff
@@ -38,62 +33,82 @@ private[mongo4cats] object helpers {
-          override def onSubscribe(s: Subscription): Unit = s.request(1)
+          private var result: T                  = _
+          private var subscription: Subscription = null
+
+          override def onNext(res: T): Unit =
+            result = res
+
+          override def onError(e: Throwable): Unit = {
+            subscription.cancel()
+            k(Left(e))
+          }
+
+          override def onComplete(): Unit = {
+            subscription.cancel()
+            k(Right(result))
+          }
+
+          override def onSubscribe(s: Subscription): Unit = {
+            subscription = s
+            subscription.request(1)
+          }
         })
       }
 
     def asyncVoid[F[_]: Async]: F[Unit] =
       Async[F].async_ { k =>
         publisher.subscribe(new Subscriber[T] {
-          override def onNext(result: T): Unit            = ()
-          override def onError(e: Throwable): Unit        = k(Left(e))
-          override def onComplete(): Unit                 = k(Right(()))
-          override def onSubscribe(s: Subscription): Unit = s.request(1)
+          private var subscription: Subscription = null
+
+          override def onNext(result: T): Unit = ()
+
+          override def onError(e: Throwable): Unit = {
+            subscription.cancel()
+            k(Left(e))
+          }
+
+          override def onComplete(): Unit = {
+            subscription.cancel()
+            k(Right(()))
+          }
+
+          override def onSubscribe(s: Subscription): Unit = {
+            subscription = s
+            subscription.request(1)
+          }
         })
       }
 
     def asyncIterable[F[_]: Async]: F[Iterable[T]] =
       Async[F].async_ { k =>
         publisher.subscribe(new Subscriber[T] {
-          private val results: ListBuffer[T]              = ListBuffer.empty[T]
-          override def onSubscribe(s: Subscription): Unit = s.request(Long.MaxValue)
-          override def onNext(result: T): Unit            = results += result
-          override def onError(e: Throwable): Unit        = k(Left(e))
-          override def onComplete(): Unit                 = k(Right(results.toList))
+          private val results: ListBuffer[T]     = ListBuffer.empty[T]
+          private var subscription: Subscription = null
+
+          override def onSubscribe(s: Subscription): Unit = {
+            subscription = s
+            subscription.request(Long.MaxValue)
+          }
+
+          override def onNext(result: T): Unit =
+            results += result
+
+          override def onError(e: Throwable): Unit = {
+            subscription.cancel()
+            k(Left(e))
+          }
+
+          override def onComplete(): Unit = {
+            subscription.cancel()
+            k(Right(results.toList))
+          }
         })
       }
-
     def stream[F[_]: Async]: Stream[F, T] =
-      mkStream(Queue.unbounded)
+      reactivestreams.fromPublisher(publisher, Int.MaxValue)
```

Comment:
I don't think we want `MaxValue` as this [won't emit until we hit that value](https://github.com/typelevel/fs2/blob/main/reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala#L169) (essentially disables backpressure and buffers everything into memory), maybe default to something like 2048? 1024?